### PR TITLE
fix: Null pointer exception when pubKeyHash is null

### LINKED ISSUE #521: pubKeyHash.toJson leads to NullPointerException
### Describe the bug

In at_client, the monitor connection fails to start with the below exception

```
WARNING|2024-02-21 10:29:34.818178|Monitor (@alice🛠)|runZonedGuarded received socket error Converting object to an encodable object failed: Instance of 'Metadata' - calling _handleError 

INFO|2024-02-21 10:29:34.818691|Monitor (@alice🛠)|socket.listen onDone called. Will destroy socket, set status stopped, call retryCallback 

INFO|2024-02-21 10:29:34.818936|Monitor (@alice🛠)|Monitor error Converting object to an encodable object failed: Instance of 'Metadata' - calling the retryCallback 
```

The reason for the above failure is because the Metadata object fails JSON Encode due to the NullPointerException. Attaching the error stack trace

```
#0      Metadata.toJson (package:at_commons/src/keystore/at_key.dart:687:60)
#1      _defaultToEncodable (dart:convert/json.dart:627:55)
#2      _JsonStringifier.writeObject (dart:convert/json.dart:787:36)
#3      _JsonStringifier.writeMap (dart:convert/json.dart:874:7)
#4      _JsonStringifier.writeJsonValue (dart:convert/json.dart:829:21)
#5      _JsonStringifier.writeObject (dart:convert/json.dart:784:9)
#6      _JsonStringStringifier.printOn (dart:convert/json.dart:982:17)
#7      _JsonStringStringifier.stringify (dart:convert/json.dart:967:5)
#8      JsonEncoder.convert (dart:convert/json.dart:345:30)
#9      JsonCodec.encode (dart:convert/json.dart:231:45)
#10     jsonEncode (dart:convert/json.dart:114:10)
#11     NotificationServiceImpl._internalNotificationCallback (package:at_client/src/service/notification_service_impl.dart:236:15)
<asynchronous suspension>
```

Below condition in `at_key.dart` is the root cause the above mentioned exceptions. The default value of `fullJson` is true. Since `fullJson` is true and the condtion is "OR", even if `pubKeyHash` is "NULL", the control still enters into IF condition and toJson() is invoked on NULL object leading to the above exceptions.

```
if (fullJson || pubKeyHash != null) {
      map[AtConstants.sharedWithPublicKeyHash] = pubKeyHash!.toJson();
    }
```

### Steps to reproduce

1. Point at_commons version to 4.0.2
2. Run the tests related to monitor connection.
3. For instance: test/enrollment_test.dart: A test to verify enrollment request returns notification in at_client functional test.

### Expected behavior

The Monitor connection should be started successfully.

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
- Fixes https://github.com/atsign-foundation/at_libraries/issues/521

**- How I did it**
- when `fullJson` is set to true (the default value), the resultant map contains all the fields. If the value is not set, null is populated. To be consistent with the other fields, changed the implementation as `pubKeyHash!.toJson()` to `pubKeyHash?.toJson()`

**- How to verify it**
- Raised a draft PR in at_client_sdk to verify the changes: https://github.com/atsign-foundation/at_client_sdk/pull/1245
- Added unit tests to assert the changes.

**- Description for the changelog**
- fix: Null pointer exception when pubKeyHash is null

### CODE REVIEW

**Review by gkc — CHANGES_REQUESTED** [review]
Please add unit tests which would have caught this issue had they existed.

**Comment by sitaram-kalluri** [comment]
> Please add unit tests which would have caught this issue had they existed.

Added unit tests and also changed the  approach in fix the bug.

when fullJson is set to true (the default value), the resultant map contains all the fields. If the value is not set, null is populated. To be consistent with the other fields, changed the implementation as `pubKeyHash!.toJson()` to `pubKeyHash?.toJson()`

File: packages/at_commons/test/at_key_test.dart [hunk 1]
```diff
@@ -987,4 +987,60 @@ void main() {
...
+      ..ttb = 100
+      ..ttr = 1
+      ..ccd = true
+      ..dataSignature = 'dummy_data_signature'
+      ..isPublic = true
+      ..isEncrypted = true
+      ..isCached = true
+      ..sharedKeyEnc = 'dummy_shared_key_enc'
+      ..pubKeyHash = PublicKeyHash('test_hash', PublicKeyHashingAlgo.sha256)
+      ..isBinary = true;
```
Comment (gkc) [inline packages/at_commons/test/at_key_test.dart]: Please extend to set all fields on the metadata and expect on all of them later

**Review by gkc — CHANGES_REQUESTED** [review]

File: packages/at_commons/test/at_key_test.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_commons/test/at_key_test.dart]: The change is implemented in https://github.com/atsign-foundation/at_libraries/pull/523/commits/8f9f23f468c445106d5f94347e2b42477b9e17f8

**Review by gkc — APPROVED** [review]

**Pull request merged**